### DISCUSSION TITLE
Embed 3D and AR into WIX website using iframe and model-viewer

### DISCUSSION THREAD

**Discussion by nemesis-ren** [google/model-viewer#3271] [Q&A]
Hi. 
I’m currently working on a Web AR experience and I can’t figure out how to use model-viewer for my website. 

**Problem**: I have a bunch of 3D models and I want to display them in a website which I’m creating for a client in WIX. 

**Looking for solutions** : 
1. How to use model-viewer to embed an iframe in the website?
2. Where do I store my 3D models(Cloud storage/database) so that I can use them in my website? 

Thanks.

**Comment by elalish** [comment]
> 1. How to use model-viewer to embed an iframe in the website?

Just any other iframe, see the src of https://modelviewer.dev/examples/twitter/generator.html for an example.

> 2. Where do I store my 3D models(Cloud storage/database) so that I can use them in my website?

Anywhere, so long as permissive CORS headers are set on the served files. Lots of people here use Firebase; you may want to search that term in the discussions.

**Reply by nemesis-ren** [reply]
I'm putting this code into WIX iFrame

<iframe src="https://cdn.glitch.global/ba7e2fea-c768-4d87-9f1c-c7f1db7e16eb/Salad.glb?v=1647436995320" style="width: 300px; height: 300px;"></iframe>

This is the model I've used - https://cdn.glitch.global/ba7e2fea-c768-4d87-9f1c-c7f1db7e16eb/Salad.glb?v=1647436995320

I'm getting blank page in iFrame. Do I need to reference model-viewer anywhere in the iFrame?

**Reply by nemesis-ren** [reply]
Basically, I'm making a webpage to display 3D model of a certain object and interaction of that object in AR. Till now I've been using Vectary, but I want to do this without vectary on my own platform. So is there any way, that I can implement this. 
![image](https://user-images.githubusercontent.com/62089287/158601913-278a351c-6868-4184-9aaa-d59875ad296a.png)

**Reply by elalish** [reply]
Yes, an iframe's src points to a webpage; in your case you'll want a webpage with model-viewer and the GLB set there. Again, look at my iframe example above.

**Reply by nemesis-ren** [reply]
![image](https://user-images.githubusercontent.com/62089287/158653726-9753ea61-24b4-4754-a202-27ac6ce56585.png)


It worked, I've used model-viewer's editor as a webpage to load the 3D view into my website, but the only problem is that it's too slow, might be because of model-viewer's database(I just dragged and dropped the model into the editor )


![image](https://user-images.githubusercontent.com/62089287/158654122-bbe631ee-7d39-43d0-b990-18710ee3122c.png)

**Reply by elalish** [reply]
Not sure what you're going for exactly, but I doubt you want the whole editor in your page. You might actually want to try linking the same twitter embed page as our iframe example link above uses; this way you can specify the model URL and other customizations using query params, and it's nice and light-weight.

**Reply by nemesis-ren** [reply]
![Screenshot 2022-03-19 130723](https://user-images.githubusercontent.com/62089287/159112563-2bdc7fc2-6dfa-4e44-9207-1c6053d3161f.png)

How do I increase the height and width of the 3D viewer? I tried putting height and width parameters in the source but it's not working.

**Reply by elalish** [reply]
Those are CSS properties, so you specify them as `style="width: 1000px; height: 1000px;"` just like for a `<div>`.

**Reply by MarcusAmazemus** [reply]
I am having a similar problem, but with respect to the files being stored. It appears to function fine if I host the model on Glitch but when I point the src to my subdomain I can't get the file to do the3D movement part first. It stays at the poster. the AR portion works, but not the interactive 3D Model part at the begining. https://www.gatewaywebar.com with the lego spaceman. Currently the src is pointing to Glitch so it functions, the problem is that I will have dozens of models with 10,000 accesses per month so I need to store the models on a subdomain.

**Reply by elalish** [reply]
@MarcusAmazemus Have you checked your JS console for errors? I'd guess this is a CORS problem: https://modelviewer.dev/docs/faq.html#entrydocs-technical-questions-cors

**Comment by ThorstenBux** [comment]
@nemesis-ren , are you after something like this? https://tripod-digital.co.nz/art/ scroll down to the 3D view.

**Reply by nemesis-ren** [reply]
Yes. With additional feature of Augmented reality.

**Reply by ThorstenBux** [reply]
yes it enables AR if the page is visited on an AR capable device. Is there a way I can reach you to discuss.

**Reply by nemesis-ren** [reply]
Yes, we can connect on Twitter to discuss this further.

**Reply by FelipeA96** [reply]
Hello @nemesis-ren i think you and i have a very similar project, i have already accomplish to succesfully load a model using iframe as @ThorstenBux suggested here and in an previous discussion (sorry i couldnt find it). Is okay if i can contact you @ThorstenBux via twitter?, i want to do the same with another model and now doesnt load the AR capabilities but the last model still loads perfectly in the same page. Thanks it has been of great help

**Reply by nemesis-ren** [reply]
Hi @FelipeA96, let's discuss on Twitter. I'll help you with whatever I can. 
My twitter handle is @Aradhye_Shubham

You can DM me. Let's help each other out.

**Reply by StillOnMyWay** [reply]
Old thread, but for Wix sites you don't need to embed them as an iframe, you can include them natively in the DOM, and get the full browser AR capability using custom elements. E.g. (my site) https://www.inthemeta.space/arr

**Reply by techseren** [reply]
@StillOnMyWay  Can you share the steps how can we directly add it to wix website

**Reply by techseren** [reply]
@StillOnMyWay  please comment the exact code snippet which you are using and where did you inserted it

**Reply by StillOnMyWay** [reply]
Because it will vary depending on the files, here's an attached walkthrough of installing it on the site. 

https://app.claap.io/threed-software/model-viewer-demo-c-HGm4RhKene-h977mp3xHk43

**Reply by StillOnMyWay** [reply]
Live demo - https://www.test-editor-x-3.wixdev-sites.org/model-viewer-two

**Comment by StillOnMyWay** [comment]
Adding to the tutorial [above](https://github.com/google/model-viewer/discussions/3271#discussioncomment-4935017), you can now also store your 3D .glb assets in Wix, and retrieve the download URL for them. This means no need to host your files elsewhere on external platforms, once they're less than 25 MB. 

There's a little code needed to get the file download URLs

https://app.claap.io/threed-software/how-to-upload-3-d-assets-to-your-wix-site-storage-c-HGm4RhKene-OUTmE-6vm_jc


```
//In any backend.jsw file
import {files} from 'wix-media.v2';

export async function fileNames(){
    return await files.listFiles(); 
}```

**Reply by AmbiguousX** [reply]
this tutorial is super helpful. how do you handle the usdz part for ios viewing in AR?

The basic astronaut example works fine, but when looking at the model viewer docs the usdz file is required for the ios src attribute of the model viewer. I'm looking to get an animated model working in AR on ios.

This works on my custom coded website, but have been struggling to get this on wix.

Thanks

**Reply by StillOnMyWay** [reply]
Thanks! Appreciate that. 

For the iOS part, I have been using the auto-generated USDZ that gets created on the fly for testing. Of course, you lose the animations that are embedded in the GLB, but I've been glad that it doesn't show up blank. 

Unfortunately Wix doesn't allow for uploads of USDZ files into storage as yet, through the method above. It's not on the accepted files list.  If you have a public URL to USDZ file ready to go, try setting it up with the ios-src attribute on your custom element, and let me know how quick view handles it.
